server/selection_test.go

```diff
@@ -0,0 +1,254 @@
+	sess1.LatencyScore = 1.1
+	sel.Complete(sess1)
+	assert.Equal(sel.Size(), 3)
+	assert.Equal(len(sel.unknownSessList), 2)
+	assert.Equal(sel.knownSessList.Len(), 1)
+
+	// Select from unknownSessList
+	sess2 := sel.Select()
+	assert.Equal(sel.Size(), 2)
+	assert.Equal(len(sel.unknownSessList), 1)
+	assert.Equal(sel.knownSessList.Len(), 1)
+
+	// Set sess2.LatencyScore to be good enough
+	sess2.LatencyScore = .9
+	sel.Complete(sess2)
+	assert.Equal(sel.Size(), 3)
+	assert.Equal(len(sel.unknownSessList), 1)
+	assert.Equal(sel.knownSessList.Len(), 2)
+
+	// Select from knownSessList
+	knownSess := sel.Select()
+	assert.Equal(sel.Size(), 2)
+	assert.Equal(len(sel.unknownSessList), 1)
+	assert.Equal(sel.knownSessList.Len(), 1)
+	assert.Equal(knownSess, sess2)
+
+	sel.Clear()
+	assert.Zero(sel.Size())
+	assert.Nil(sel.unknownSessList)
+	assert.Zero(sel.knownSessList.Len())
+	assert.Nil(sel.stakeRdr)
+}
+
+func TestMinLSSelector_UnknownSessListSelect_Errors(t *testing.T) {
+	assert := assert.New(t)
+
+	stakeRdr := newStubStakeReader()
+	sel := NewMinLSSelector(stakeRdr, 1.0)
+
+	sel.Add(
+		[]*BroadcastSession{
+			&BroadcastSession{
+				OrchestratorInfo: &net.OrchestratorInfo{
+					TicketParams: &net.TicketParams{Recipient: []byte("foo")},
+				},
+			},
+		},
+	)
+
+	// Test error when reading stake
+	stakeRdr.err = errors.New("Stakes error")
+	errorLogsBefore := glog.Stats.Error.Lines()
+	assert.Nil(sel.unknownSessListSelect())
+	errorLogsAfter := glog.Stats.Error.Lines()
+	assert.Equal(int64(1), errorLogsAfter-errorLogsBefore)
+}
+
+func TestMinLSSelector_UnknownSessListSelect_UniqueWeights(t *testing.T) {
+	stakeRdr := newStubStakeReader()
+	sel := NewMinLSSelector(stakeRdr, 1.0)
+
+	sessions := make([]*BroadcastSession, 10)
+	stakes := make([]*big.Int, 10)
+	stakeMap := make(map[ethcommon.Address]*big.Int)
+	for i := 0; i < 10; i++ {
+		addr := ethcommon.BytesToAddress([]byte(strconv.Itoa(i)))
+		stake := big.NewInt(int64(1000 * i))
+
+		sessions[i] = &BroadcastSession{
+			OrchestratorInfo: &net.OrchestratorInfo{
+				TicketParams: &net.TicketParams{Recipient: addr.Bytes()},
+			},
+		}
+		stakes[i] = stake
+		stakeMap[addr] = stake
+	}
+
+	stakeRdr.SetStakes(stakeMap)
+	sel.Add(sessions)
+
+	// Run unknownSessListSelect() x100000 and record # of times a session weight is selected
+	// Each session has a unique stake weight so we will record the # of selections per stake weight
+	stakeCount := make(map[*big.Int]int)
+	for i := 0; i < 100000; i++ {
+		sess := sel.unknownSessListSelect()
+		addr := ethcommon.BytesToAddress(sess.OrchestratorInfo.TicketParams.Recipient)
+		stake := stakeMap[addr]
+		stakeCount[stake]++
+
+		// Call Add() to add the session back to unknownSessList
+		sel.Add([]*BroadcastSession{sess})
+	}
+
+	sort.Slice(stakes, func(i, j int) bool {
+		return stakes[i].Cmp(stakes[j]) < 0
+	})
+
+	// Check that higher stake weight sessions are selected more often than lower stake weight sessions
```

Comment:
I guess technically we'd want to check the ratio of selection to ensure the weighing works as expected, but this is probably OK as well.

Reply:
Fixed in 09df6bc - added a test to check that the selection count ratio matches up with the stake weight ratio.